### PR TITLE
[Kernel] Fuse Ming streaming ISTFT overlap-add and normalization

### PR DESCRIPTION
**Pull request by yashkgp**
## Purpose

Implements the kernel experiment in #7337. Ming's `padding="same"` ISTFT currently applies a window, folds samples and the squared window separately, updates streaming tails, and normalizes. This adds one Triton kernel for that post-FFT work while retaining cuFFT and the envelope-underflow check.

The fast path covers NVIDIA inference with `n_fft == win_length == 4 * hop_length`, complex64 spectra, and FP32/BF16/FP16 windows. It preserves native rounding boundaries, unnormalized carry state, and first/middle/final trimming. The original implementation remains as `forward_native` for reference, unsupported inputs, and autograd.

**The operator improves, but a consistent decoder-level latency benefit has not been established.** This is a performance feature experiment, not a correctness fix. The measured decoder changes are within run-to-run variation; the value criterion in #7337 remains open. No serving throughput, first-audio latency, CUDA Graph, or perceptual-quality improvement is claimed.

Related #4060 retains the two folds and changes graph-capture behavior; #4129 covers broader VAE streaming work. This change specifically fuses post-FFT reconstruction and keeps validation. All-state searches and related Ming diff inspection found no exact duplicate; focused searches were refreshed on 2026-09-09.

## Test Plan

One NVIDIA A10G (23,028 MiB reported VRAM), Ubuntu 22.04, driver 595.91.07, Python 3.12.14, PyTorch 2.13.0+cu130, CUDA runtime 13.0, Triton 3.7.1, transformers 5.14.1.

**vLLM Version:** 0.29.0

**vLLM-Omni Commit:** baseline `6fb7b0a05e3758e41e3871804598cea18a0b42fb`; implementation `979f43c31a7cfff76f04dd6b5954f0d2f8c22d06`.

From the repository root in the prepared environment:

```bash
python -m pytest tests/model_executor/models/ming_tts/test_fused_istft.py --run-level core_model -q
compute-sanitizer --tool memcheck --error-exitcode 1 python -m pytest tests/model_executor/models/ming_tts/test_fused_istft.py --run-level core_model -q
PYTHONPATH=. python benchmarks/kernels/benchmark_ming_istft.py --compile --decoder --warmup 20 --iterations 100 --decoder-iterations 100
pre-commit run --files vllm_omni/model_executor/models/common/ming/audio_dsp.py vllm_omni/model_executor/models/common/ming/fused_istft.py tests/model_executor/models/ming_tts/test_fused_istft.py benchmarks/kernels/benchmark_ming_istft.py
```

Tests cover both production hops (320/882), a small geometry, all three window dtypes, multiple batches, strided spectra and native carry views, five chunk modes, irregular streaming splits, independent requests, silence, invalid envelopes, unsupported input fallback, and CPU/CUDA autograd. FP32 streaming is also checked against whole-signal reconstruction. Low-precision streaming is compared to the identical native chunk sequence because native carry rounding already differs from one-shot execution.

## Test Result

- Focused suite: **195 passed, 15 warnings in 10.10s**, using the normal repository pytest configuration.
- Compute Sanitizer memcheck: **195 passed, 15 warnings in 11.86s; ERROR SUMMARY: 0 errors**.
- All applicable pre-commit checks pass. Full repository precheck and independent code/benchmark review completed; a ROCm test-gating finding was fixed and CUDA autograd coverage added.
- Maximum absolute error in the reported operator and pretrained-decoder comparisons: **3.73e-9**. The fused parity gate is `atol=5e-7, rtol=1e-5`, including both state buffers.
- The broader Ming shared-module suite could not collect with this environment: an existing serving import requires `vllm.entrypoints.launcher`, absent from installed vLLM 0.29.0. No full serving test is claimed.

The tables below are warmed, synchronized **wall-clock** latency, including Python overhead and the retained envelope check. Operator cases use batch 1, hop 882, BF16 windows and a nonempty streaming carry; each has 20 warmups and 100 measured iterations.

| Scope / spectral frames | Native p50 (ms) | Fused p50 (ms) | Latency reduction |
| --- | ---: | ---: | ---: |
| Post-FFT / 16 | 0.2322 | 0.1176 | 49.4% |
| Post-FFT / 100 | 0.2402 | 0.1262 | 47.5% |
| Post-FFT / 400 | 0.2410 | 0.1271 | 47.3% |
| Full ISTFT / 16 | 0.2958 | 0.2044 | 30.9% |
| Full ISTFT / 100 | 0.3069 | 0.2073 | 32.5% |
| Full ISTFT / 400 | 0.3068 | 0.2179 | 29.0% |

For full ISTFT / 100 frames, standard deviation is 0.0120 ms native and 0.0109 ms fused. Extra peak allocation is 4.0459 / 4.0391 MiB; isolating post-FFT work gives 4.0386 / 0.6060 MiB. This is incremental allocated memory, not total device utilization. The benchmark emits dispersion and memory for every case.

The script also measures `torch.compile(native)`. FP32 comparisons pass the parity gate; BF16 compiled output differs from native by roughly 2–3e-4 and fails it. Compiled timings are reported with that flag and are not treated as a numerically equivalent BF16 baseline.

Decoder measurement uses the actual pretrained `inclusionAI/Ming-omni-tts-0.5B` AudioVAE decoder, pinned to revision `9154772e7fbc585907b6237e3190790676f28975`, BF16/SDPA, with fixed **synthetic** latent chunks `[4, 25, 25, 5]`. Every invocation resets cache/state and flushes the last chunk. Each round has 20 warmups and 100 measurements; the native/fused order alternates.

| Decoder round | Native p50 (ms) | Fused p50 (ms) | Latency reduction |
| --- | ---: | ---: | ---: |
| 1 | 89.6255 | 89.9965 | -0.41% |
| 2 | 89.4831 | 88.9954 | 0.54% |
| 3 | 90.4878 | 90.1280 | 0.40% |

Within-round standard deviations span 0.39–1.25 ms. Extra peak allocation is 14.4883 MiB native and 11.8022 MiB fused. These results support the operator improvement and reduced temporary allocation, but do not resolve a decoder latency gain. Full text-to-speech serving and representative real latent traces remain untested.

AI assistance: Codex researched the proposal, implemented the kernel, authored tests/benchmarks, ran validation, wrote this description, and performed automated reviews. Human contributor and maintainer review remain outstanding.

### CODE REVIEW

**Comment by yashkgp** [comment]
Self-review of `979f43c31` (Codex-assisted):

- Checked the four-file diff for bounded kernel loads/stores, native accumulation order and FP32/BF16/FP16 rounding, first/middle/final trimming, and unnormalized carry state without input-buffer mutation.
- Checked native fallback for unsupported inputs and autograd, retained envelope-underflow validation, and NVIDIA/Triton test gating. The earlier ROCm test-gating finding was fixed, and CUDA backward coverage was added.
- A10G validation: 195 focused tests passed normally and under Compute Sanitizer, with zero memory errors. Applicable local pre-commit checks pass; GitHub wheel builds, pre-commit, docs and DCO now pass too.
- Checked that performance claims match the evidence: full ISTFT latency is 29–32% lower, but the three decoder rounds do not establish a consistent decoder latency gain. Full serving is untested, and the broader shared-module suite could not collect with installed vLLM 0.29.0 because `vllm.entrypoints.launcher` is absent.

These checks were performed by Codex, including the automated code review and GPU validation; they do not represent independent human review. No additional code change was needed for this comment.

**Comment by yashkgp** [comment]
Please review @vllm-omni-review-bot

**Review by linyueqian — APPROVED** [review]
Approving. The dispatch is structured the way a kernel substitution should be: `forward` only takes the fused path when padding is `same`, the input is CUDA, grad is off, and `fused_istft_supported` agrees, and `forward_native` survives untouched as the fallback rather than being replaced. That means the risky path is opt-in per call and the reference implementation stays available for comparison instead of becoming folklore.

What makes this reviewable rather than a matter of trust is the test file. `test_fused_istft_matches_native` sweeps hop, window dtype, batch and frame count across offline, streaming, last-chunk and first-and-last modes, and compares against `forward_native` at `atol=5e-7`. The part worth singling out is that it monkeypatches `forward_native` to raise before exercising the fused path, so the test cannot pass by quietly falling back, which is the usual way a fused-kernel equivalence test rots. `test_fused_istft_streaming_state` then walks chunked streaming against native chunk by chunk and also checks the concatenation against the offline native result, which is the property that actually matters for overlap-add: per-chunk agreement does not imply the seam is right. Silence, invalid envelope and the autograd fallback are covered too.

I considered running an end-to-end synthesis with ASR on this one, since it changes audio math and that is normally where a transcript earns its keep. It would have been the weaker evidence here: a transcript establishes intelligibility, while these tests establish agreement with the existing implementation to 5e-7 across the streaming seam. Reserving that for a change where equivalence is not already pinned.

CI: `buildkite/vllm-omni` green at this head. The informational lanes carry their standing reds, covered by `ci_approval_override_head`.

Validation: static review of the four-file author diff from the GitHub file list, the dispatch guard in `audio_dsp.py`, and the equivalence, streaming-seam, edge-case and autograd tests. Fork head, so the branch was not executed.
